File: core/src/main/scala/diplomacy/board/Board.scala
package diplomacy.board

final case class Board[State_, Power_ <: Power, MilitaryBranch_ <: MilitaryBranch, UnitStatus_, ProvinceStatus_](
  map: DiplomacyMap[Power_, MilitaryBranch_],
  state: State_, units: Set[DiplomacyUnit[Power_, MilitaryBranch_]],
  occupation: Map[Power_, Set[Province[Power_]]],
  unitStatuses: Map[DiplomacyUnit[Power_, MilitaryBranch_], UnitStatus_],
  provinceStatuses: Map[Province[Power_], ProvinceStatus_]
) {
  val numberOfSupplyCenters: Map[Power_, Int] = this.occupation map {
    case (power, provinces) => power -> provinces.count(province => province.isSupplyCenter)
  }

  override def toString: String = this.state.toString
}

object Board {
  trait TypeHelper extends DiplomacyMap.TypeHelper {
    type State
    type UnitStatus
    type ProvinceStatus
    type Board = diplomacy.board.Board[State, Power, MilitaryBranch, UnitStatus, ProvinceStatus]
  }
}

File: src/board/board.ts
import { DiplomacyMap } from "./diplomacy-map"
import { Unit } from "./unit"
import { Province } from "./province"

export class Board<Power, MilitaryBranch, State, UnitStatus, ProvinceStatus> {
  /**
   * The units that are in this board
   */
  units: Set<Unit<Power, MilitaryBranch>>
  /**
   * The state of each unit (e.g., the unit was dislodged)
   */
  unitStatuses: Map<Unit<Power, MilitaryBranch>, UnitStatus>
  /**
   * The state of each province (e.g., standoff was occurred, this province is occupied by X)
   */
  provinceStatuses: Map<Province<Power>, ProvinceStatus>
  /**
   * @param map
   * @param state
   * @param units The units that are in this board
   * @param unitStatuses The state of each unit (e.g., the unit was dislodged)
   * @param provinceStatuses
   *   The state of each province (e.g., standoff was occurred, this province is occupied by X)
   */
  constructor (
    public map: DiplomacyMap<Power, MilitaryBranch>, public state: State,
    units: Set<Unit<Power, MilitaryBranch>> | Array<Unit<Power, MilitaryBranch>>,
    unitStatuses: Map<Unit<Power, MilitaryBranch>, UnitStatus> | Array<[Unit<Power, MilitaryBranch>, UnitStatus]>,
    provinceStatuses: Map<Province<Power>, ProvinceStatus> | Array<[Province<Power>, ProvinceStatus]>
  ) {
    this.units = new Set([...units])
    this.unitStatuses = new Map([...unitStatuses])
    this.provinceStatuses = new Map([...provinceStatuses])
  }
}
